File: packages/agent/src/service.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { execFileSync } from 'node:child_process'

export const SERVICE_NAME = 'webmux-agent'

export interface ManagedRelease {
  cliPath: string
  releaseDir: string
}

export interface RenderServiceUnitOptions {
  agentName: string
  homeDir: string
  nodePath: string
  cliPath: string
  autoUpgrade: boolean
  pathEnv: string
}

export interface InstallServiceOptions {
  agentName: string
  packageName: string
  version: string
  autoUpgrade: boolean
  homeDir?: string
}

export interface UpgradeServiceOptions {
  agentName: string
  packageName: string
  version: string
  autoUpgrade?: boolean
  homeDir?: string
}

export interface InstalledServiceConfig {
  autoUpgrade: boolean
  version: string | null
}

export function renderServiceUnit(options: RenderServiceUnitOptions): string {
  return `[Unit]
Description=Webmux Agent (${options.agentName})
After=network-online.target
Wants=network-online.target

[Service]
Type=simple
ExecStart=${options.nodePath} ${options.cliPath} start
Restart=always
RestartSec=10
Environment=WEBMUX_AGENT_SERVICE=1
Environment=WEBMUX_AGENT_AUTO_UPGRADE=${options.autoUpgrade ? '1' : '0'}
Environment=WEBMUX_AGENT_NAME=${options.agentName}
Environment=HOME=${options.homeDir}
Environment=PATH=${options.pathEnv}
WorkingDirectory=${options.homeDir}

[Install]
WantedBy=default.target
`
}

export function installService(options: InstallServiceOptions): void {
  const homeDir = options.homeDir ?? os.homedir()
  const autoUpgrade = options.autoUpgrade
  const release = installManagedRelease({
    packageName: options.packageName,
    version: options.version,
    homeDir,
  })

  writeServiceUnit({
    agentName: options.agentName,
    autoUpgrade,
    cliPath: release.cliPath,
    homeDir,
  })

  runSystemctl(['--user', 'daemon-reload'])
  runSystemctl(['--user', 'enable', SERVICE_NAME])
  runSystemctl(['--user', 'restart', SERVICE_NAME])
  runCommand('loginctl', ['enable-linger', os.userInfo().username])
}

export function upgradeService(options: UpgradeServiceOptions): void {
  const homeDir = options.homeDir ?? os.homedir()
  const installedConfig = readInstalledServiceConfig(homeDir)
  const autoUpgrade = options.autoUpgrade ?? installedConfig?.autoUpgrade ?? true
  const release = installManagedRelease({
    packageName: options.packageName,
    version: options.version,
    homeDir,
  })

  writeServiceUnit({
    agentName: options.agentName,
    autoUpgrade,
    cliPath: release.cliPath,
    homeDir,
  })

  runSystemctl(['--user', 'daemon-reload'])
  runSystemctl(['--user', 'restart', SERVICE_NAME])
}

export function uninstallService(homeDir = os.homedir()): void {
  const unitPath = servicePath(homeDir)

  try {
    runSystemctl(['--user', 'stop', SERVICE_NAME])
  } catch {
    // Ignore if the service is not running.
  }

  try {
    runSystemctl(['--user', 'disable', SERVICE_NAME])
  } catch {
    // Ignore if the service is not installed.
  }

  if (fs.existsSync(unitPath)) {
    fs.unlinkSync(unitPath)
  }

  try {
    runSystemctl(['--user', 'daemon-reload'])
  } catch {
    // Ignore reload errors during cleanup.
  }
}

export function readInstalledServiceConfig(homeDir = os.homedir()): InstalledServiceConfig | null {
  const unitPath = servicePath(homeDir)
  if (!fs.existsSync(unitPath)) {
    return null
  }

  const unit = fs.readFileSync(unitPath, 'utf-8')
  const autoUpgradeMatch = unit.match(/^Environment=WEBMUX_AGENT_AUTO_UPGRADE=(\d)$/m)
  const versionMatch = unit.match(/\/releases\/([^/\s]+)\/node_modules\//)

  return {
    autoUpgrade: autoUpgradeMatch?.[1] !== '0',
    version: versionMatch?.[1] ?? null,
  }
}

export function servicePath(homeDir = os.homedir()): string {
  return path.join(homeDir, '.config', 'systemd', 'user', `${SERVICE_NAME}.service`)
}

function writeServiceUnit(options: {
  agentName: string
  autoUpgrade: boolean
  cliPath: string
  homeDir: string
}): void {
  const serviceDir = path.dirname(servicePath(options.homeDir))
  fs.mkdirSync(serviceDir, { recursive: true })
  fs.writeFileSync(
    servicePath(options.homeDir),
    renderServiceUnit({
      agentName: options.agentName,
      autoUpgrade: options.autoUpgrade,
      cliPath: options.cliPath,
      homeDir: options.homeDir,
      nodePath: findBinary('node') ?? process.execPath,
      pathEnv: process.env.PATH ?? '',
    }),
  )
}

function installManagedRelease(options: {
  packageName: string
  version: string
  homeDir: string
}): ManagedRelease {
  const releaseDir = path.join(options.homeDir, '.webmux', 'releases', options.version)
  const cliPath = path.join(
    releaseDir,
    'node_modules',
    ...options.packageName.split('/'),
    'dist',
    'cli.js',
  )

  if (fs.existsSync(cliPath)) {
    return { cliPath, releaseDir }
  }

  fs.mkdirSync(releaseDir, { recursive: true })
  ensureRuntimePackageJson(releaseDir)

  const packageManager = findBinary('pnpm') ? 'pnpm' : 'npm'
  if (packageManager === 'pnpm') {
    runCommand('pnpm', ['add', '--dir', releaseDir, `${options.packageName}@${options.version}`])
  } else {
    if (!findBinary('npm')) {
      throw new Error('Cannot find pnpm or npm. Install one package manager before installing the service.')
    }
    runCommand('npm', ['install', '--omit=dev', `${options.packageName}@${options.version}`], releaseDir)
  }

  if (!fs.existsSync(cliPath)) {
    throw new Error(`Managed release did not produce a CLI at ${cliPath}`)
  }

  return { cliPath, releaseDir }
}

function ensureRuntimePackageJson(releaseDir: string): void {
  const packageJsonPath = path.join(releaseDir, 'package.json')
  if (fs.existsSync(packageJsonPath)) {
    return
  }

  fs.writeFileSync(
    packageJsonPath,
    JSON.stringify({
      name: 'webmux-agent-runtime',
      private: true,
    }, null, 2) + '\n',
  )
}

function runSystemctl(args: string[]): void {
  runCommand('systemctl', args)
}

function runCommand(command: string, args: string[], cwd?: string): void {
  execFileSync(command, args, {
    cwd,
    stdio: 'inherit',
  })
}

function findBinary(name: string): string | null {
  try {
    return execFileSync('which', [name], { encoding: 'utf-8' }).trim()
  } catch {
    return null
  }
}

File: rust/crates/webmux-agent/src/service.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

pub const SERVICE_NAME: &str = "webmux-node";

/// Render a systemd unit file.
pub fn render_service_unit(
    agent_name: &str,
    home_dir: &str,
    cli_path: &str,
    auto_upgrade: bool,
    path_env: &str,
) -> String {
    let auto_upgrade_val = if auto_upgrade { "1" } else { "0" };
    format!(
        r#"[Unit]
Description=Webmux Node ({agent_name})
After=network-online.target
Wants=network-online.target

[Service]
Type=simple
ExecStart={cli_path} start
Restart=always
RestartSec=10
Environment=WEBMUX_AGENT_SERVICE=1
Environment=WEBMUX_AGENT_AUTO_UPGRADE={auto_upgrade_val}
Environment=WEBMUX_AGENT_NAME={agent_name}
Environment=HOME={home_dir}
Environment=PATH={path_env}
WorkingDirectory={home_dir}

[Install]
WantedBy=default.target
"#
    )
}

pub struct InstallServiceOptions {
    pub agent_name: String,
    pub auto_upgrade: bool,
}

/// Install and start the agent as a systemd user service. The Rust binary is
/// the service executable itself (no Node.js or managed releases needed).
pub fn install_service(options: &InstallServiceOptions) -> Result<(), String> {
    let home_dir = dirs::home_dir()
        .ok_or_else(|| "cannot determine home directory".to_string())?;
    let home_str = home_dir.to_string_lossy().to_string();

    let cli_path = std::env::current_exe()
        .map_err(|e| format!("failed to determine current executable path: {e}"))?
        .to_string_lossy()
        .to_string();

    let path_env = std::env::var("PATH").unwrap_or_default();

    let unit = render_service_unit(
        &options.agent_name,
        &home_str,
        &cli_path,
        options.auto_upgrade,
        &path_env,
    );

    let unit_path = service_path(&home_str);
    let unit_dir = PathBuf::from(&unit_path)
        .parent()
        .map(|p| p.to_path_buf())
        .ok_or_else(|| "cannot determine service directory".to_string())?;

    fs::create_dir_all(&unit_dir)
        .map_err(|e| format!("failed to create service directory: {e}"))?;
    fs::write(&unit_path, &unit)
        .map_err(|e| format!("failed to write service unit: {e}"))?;

    run_systemctl(&["--user", "daemon-reload"])?;
    run_systemctl(&["--user", "enable", SERVICE_NAME])?;
    run_systemctl(&["--user", "restart", SERVICE_NAME])?;

    // Enable lingering so the service starts at boot without a login session
    let username = whoami();
    if let Some(user) = username {
        let _ = run_command("loginctl", &["enable-linger", &user]);
    }

    Ok(())
}

/// Stop and remove the systemd user service.
pub fn uninstall_service() -> Result<(), String> {
    let home_dir = dirs::home_dir()
        .ok_or_else(|| "cannot determine home directory".to_string())?;
    let home_str = home_dir.to_string_lossy().to_string();
    let unit_path = service_path(&home_str);

    let _ = run_systemctl(&["--user", "stop", SERVICE_NAME]);
    let _ = run_systemctl(&["--user", "disable", SERVICE_NAME]);

    if PathBuf::from(&unit_path).exists() {
        fs::remove_file(&unit_path)
            .map_err(|e| format!("failed to remove service file: {e}"))?;
    }

    let _ = run_systemctl(&["--user", "daemon-reload"]);
    Ok(())
}

/// Read installed service configuration from the unit file.
pub struct InstalledServiceConfig {
    pub auto_upgrade: bool,
}

pub fn read_installed_service_config() -> Option<InstalledServiceConfig> {
    let home_dir = dirs::home_dir()?;
    let home_str = home_dir.to_string_lossy().to_string();
    let unit_path = service_path(&home_str);

    let unit = fs::read_to_string(&unit_path).ok()?;

    let auto_upgrade = if let Some(pos) = unit.find("WEBMUX_AGENT_AUTO_UPGRADE=") {
        let rest = &unit[pos + "WEBMUX_AGENT_AUTO_UPGRADE=".len()..];
        !rest.starts_with('0')
    } else {
        true
    };

    Some(InstalledServiceConfig { auto_upgrade })
}

/// Show systemd service status (runs `systemctl --user status webmux-agent` with
/// inherited stdio).
pub fn show_service_status() {
    let _ = Command::new("systemctl")
        .args(["--user", "status", SERVICE_NAME])
        .status();
}

/// Check whether the service is active. Returns the status string (e.g. "active")
/// or `None` if the service is not installed.
pub fn is_service_active() -> Option<String> {
    let output = Command::new("systemctl")
        .args(["--user", "is-active", SERVICE_NAME])
        .output()
        .ok()?;
    let status = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if output.status.success() {
        Some(status)
    } else {
        None
    }
}

pub fn service_path(home_dir: &str) -> String {
    let mut p = PathBuf::from(home_dir);
    p.push(".config");
    p.push("systemd");
    p.push("user");
    p.push(format!("{SERVICE_NAME}.service"));
    p.to_string_lossy().to_string()
}

fn run_systemctl(args: &[&str]) -> Result<(), String> {
    run_command("systemctl", args)
}

fn run_command(cmd: &str, args: &[&str]) -> Result<(), String> {
    let status = Command::new(cmd)
        .args(args)
        .status()
        .map_err(|e| format!("failed to execute {cmd}: {e}"))?;

    if !status.success() {
        return Err(format!("{cmd} exited with status {status}"));
    }
    Ok(())
}

fn whoami() -> Option<String> {
    let output = Command::new("whoami").output().ok()?;
    if output.status.success() {
        Some(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        None
    }
}
